qcp/matrix.py

```diff
@@ -0,0 +1,150 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from __future__ import annotations
+from abc import ABC
+from typing import Union
+from ._types import SCALARS, VECTOR, MATRIX
+
+class Matrix(ABC):
+    """
+    Method stubs for an immutable implementation of a matrix.
+    """
+
+    def __init__(self, state: MATRIX):
+        self.state = state
+        self.conjugated = False
+    
+    def dim(self):
+    #Return the dimension of matrix, in (row,col) tuple
+    #__len__ cannot return tuple, only integer
+        if not isinstance(self.state[0], list):
+            return (len(self.state),1)
+        else:
+            return (len(self.state),len(self.state[0]))     
+
+    def __getitem__(self, i: int) -> VECTOR:
+        pass
+
+    def __setitem__(self, i: int, v: SCALARS):
+        pass
+
+    def get_state(self) -> MATRIX:
+        s = self.state
+        return s
+
+    def set_state(self, s: MATRIX):
+        self.state = s
+        return
+
+    @classmethod
+    def zeros(cls,nrow, ncol=1):
+        #Create zero matrix with dimension (nrow,ncol)
+        #Class method used to handle the creation of new object
+        self = cls([])
+        temp = []
+        for i in range(nrow):
+            for j in range(ncol):
+               temp.append(0)
+            self.state.append(temp)
+            temp = []    
+        return self
+
+    def __add__(self, other: Matrix) -> Matrix:
+        #Addition between two matrices with same dimension
+        assert(isinstance(other, type(self)))
+        assert self.dim() == other.dim(),'Cannot add matrices with different dimensions'
+        nrow,ncol = self.dim()
+        sum = self
+        for i in range(nrow):
+            for j in range(ncol):
+                sum[i][j] = self.state[i][j] + other.state[i][j]
+        return sum
+        
+
+    def __sub__(self, other: Matrix) -> Matrix:
+        #Subtraction between two matrices with same dimension
+        assert (isinstance(other, type(self)))
+        assert self.dim() == other.dim(),'Cannot add matrices with different dimensions'
+        nrow,ncol = self.dim()
+        subtraction = self.state
+        for i in range(nrow):
+            for j in range(ncol):
+                subtraction[i][j] = self.state[i][j] - other.state[i][j]
+        return subtraction
+
+    def columns(self) -> MATRIX:
+        pass
+
+    def __mul__(self, other):
+        #Multiplication
+        #Check if the dimensions of the two matrices are compatible
+        assert self.dim()[1] == other.dim()[0],'Cannot add matrices with different dimensions'
+        #If the other input is an integer, multiply directly
+        if isinstance(other, int):
+            nrow,ncol = self.dim()
+            for i in range(nrow):
+              for j in range(ncol):
+                self.state[i][j] = self.state[i][j] * other
+            return self                 
```

Comment:
I think ideally instead of returning `self` here, we should be creating a new Matrix variable, otherwise we should be consistent about the mutability of the `Matrix` class, and the way the dot producting works should be modified.